File: app/src/main/java/com/example/sysinfo/ui/main/SectionsPagerAdapter.java
package com.example.sysinfo.ui.main;

import android.content.Context;

import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;

import com.example.sysinfo.R;
import com.example.sysinfo.fragments.CPU;
import com.example.sysinfo.fragments.Dashboard;
import com.example.sysinfo.fragments.DeviceFragment;
import com.example.sysinfo.fragments.SystemFrag;


/**
 * A [FragmentPagerAdapter] that returns a fragment corresponding to
 * one of the sections/tabs/pages.
 */

public class SectionsPagerAdapter extends FragmentPagerAdapter {

    @StringRes
    private static final int[] TAB_TITLES = new int[]{R.string.tab_text_1, R.string.tab_text_2,R.string.tab_text_3,R.string.tab_text_4};
    private final Context mContext;

    public SectionsPagerAdapter(Context context, FragmentManager fm) {
        super(fm);
        mContext = context;
    }

    @Override
    public Fragment getItem(int position) {
        Fragment fragment = null;
        switch (position){
            case 0:
                fragment = new Dashboard();
                break;
            case 1:
                fragment = new DeviceFragment();
                break;
            case 2:
                fragment = new SystemFrag();
                break;
            case 3:
                fragment = new CPU();
                break;
        }
        return fragment;
    }


    @Nullable
    @Override
    public CharSequence getPageTitle(int position) {
        return mContext.getResources().getString(TAB_TITLES[position]);
    }

    @Override
    public int getCount() {
        // Show 4 total pages.
        return 4;
    }
}

File: app/src/main/java/com/kl3jvi/sysinfo/ui/main/SectionsPagerAdapter.kt
package com.kl3jvi.sysinfo.ui.main

import android.content.Context
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.example.sysinfo.R
import com.kl3jvi.sysinfo.view.fragments.BatteryFragment
import com.kl3jvi.sysinfo.view.fragments.CPU
import com.kl3jvi.sysinfo.view.fragments.Dashboard
import com.kl3jvi.sysinfo.view.fragments.DeviceFragment
import com.kl3jvi.sysinfo.view.fragments.ScreenFragment
import com.kl3jvi.sysinfo.view.fragments.StorageFragment
import com.kl3jvi.sysinfo.view.fragments.SystemFrag

class SectionsPagerAdapter(
    private val context: Context,
    fm: FragmentActivity
) : FragmentStateAdapter(fm) {

    enum class TabType(val titleResId: Int, val fragment: () -> Fragment) {
        DASHBOARD(R.string.tab_text_1, ::Dashboard),
        DEVICE(R.string.tab_text_2, ::DeviceFragment),
        SYSTEM(R.string.tab_text_3, ::SystemFrag),
        STORAGE(R.string.tab_text_4, ::StorageFragment),
        CPU(R.string.tab_text_5, { CPU() }),
        BATTERY(R.string.tab_text_6, ::BatteryFragment),
        SCREEN(R.string.tab_text_7, ::ScreenFragment);

        companion object {
            fun fromPosition(position: Int) = values().getOrElse(position) {
                throw IllegalStateException("Invalid position: $position")
            }
        }
    }

    override fun getItemCount(): Int = TabType.values().size

    override fun createFragment(position: Int): Fragment {
        return TabType.fromPosition(position).fragment.invoke()
    }

    fun getPageTitle(position: Int): CharSequence {
        return context.resources.getString(TabType.fromPosition(position).titleResId)
    }
}